### PR TITLE
Backport of access to tauID and electron discrimination from miniAOD

### PR DESCRIPTION
**Pull request by roger-wolf**
Dear colleagues, 
this is a 1:1 backport of the work we have invested last autumn to provide access to tau-ID and electron discrimination on miniAOD in 81X. The relevant PRs have been:

 * https://github.com/cms-sw/cmssw/pull/15731
 * https://github.com/cms-sw/cmssw/pull/15747
 * https://github.com/cms-sw/cmssw/pull/15758
 * https://github.com/cms-sw/cmssw/pull/15995
 * https://github.com/cms-sw/cmssw/pull/16030

By that time we had no chance to get these developments backported in time to the 80X release series and had announced that we would like to do this once we are set. We have now prepared the first case where tau-ID and electron-discrimination have indeed been produced and can be applied to analyses based only on miniAOD information! This has been provided for the upcoming analyses of this year. It is consequently done in 80X. 

The backport introduces necessary additional contents in a few tau-specific DataFormats. No existing contents or members are changed or removed. In standard workflows there is no access to the newly added data members, so the backport does not change standard reconstruction behaviour. It only allows that existing training files may also be accessed from miniAOD in addition to the unchanged standard access from AOD. This project is of great importance for the flexibility of tau analyses in 2016. 

Since only information is added we do not expect any change of the default reconstruction behaviour (see discussion of 81X pull requests). We have made the standard tests that we usually do in preparation of our pull requests to verify that this is the case. The additional features have also been successfully tested with unit-tests that had been introduced by the time of the original pull requests to 81X (and backported accordingly). In addition the whole infrastructure has already been used for the production and extensive validation of new MVA weight files, and their application. A detailed report is planed by Alexander Nehrkorn for the TauPOG meeting during the upcoming CMS week. 

A small set of modifications on top of this pull request (which is a plain backport as said) would follow to give official access to the new weight files, most probably also related to a new Global Tag request. We would highly appreciate if these modifications could make it into a new 80X analysis release if possible. I'm happy to give a report or just to drop a few words of further clarification during the RECO/AT meeting tomorrow. You can put me on the agenda and let me know in case. 

Cheers,
Roger

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@roger-wolf 
you mention #15758 in the PR description. This is an 80X PR. 
I'm guessing that you had in mind #15778

**Comment by roger-wolf** [comment]
Hi Slava,

you are right -- https://github.com/cms-sw/cmssw/pull/15758 is already a backport. So I must have mixed this in by incidence. Please note though that we would like to add one single variable (phiAtEcalEntrance_) to TauPFEssentials.h in PatCandidates if possible. This variable was missing for the backport of https://github.com/cms-sw/cmssw/pull/15758 and the fix did not appear in time by that time. Since this is only an additional variable in the PAT event contents that is default filled 0 unless the new tauID sequence is run I hope that adding this member is not impossible. 

Cheers,
Roger

**Comment by roger-wolf** [comment]
Hi all, 

all tests passed all questions answered as far as there were any. Any reasons not to sign this off so that we could continue with a few more commits on top of this?

Thanx a lot!

Cheers,
Roger

**Comment by slava77** [comment]
Hi Roger,
I haven't checked the backport yet. That's why it's not signed.

The PR description states "this is a 1:1 backport".
I made a quick attempt to add up to the +3,083 lines in this PR  and was not successful:

```
 Extension of miniAOD event content to be able to run tau mva isolation discriminator on miniAOD input #15731 
https://github.com/cms-sw/cmssw/pull/15731	in CMSSW_8_1_X
+81 −43	 in 8
 Extension of PAT TauPFEssential event content to be able to run anti-electron discriminators in miniAOD input #15747 
https://github.com/cms-sw/cmssw/pull/15747	in CMSSW_8_1_X
+116 −41	 in 5

 Extension of miniAOD event content to be able to run tau mva isolation discriminator on miniAOD input #15758 
https://github.com/cms-sw/cmssw/pull/15758	in CMSSW_8_0_X from Sep 8, 2016 ??? 
+145 −22	 in 6
.. this does not belong in here as a reference

 Adding missing variable to be able to run the anti-electron discriminator on present trainings #15995 
https://github.com/cms-sw/cmssw/pull/15995	in CMSSW_8_1_X
+23 −10	in 7
 Extension of infrastructure to readout tau discriminators from miniAOD to anti-electron discriminator #16030 
https://github.com/cms-sw/cmssw/pull/16030	in CMSSW_8_1_X
+607 −24	in 6

.. this clearly doesn't add up
.. there was also
 Infrastructure to create the actual ValueMap pat::Tau-discriminator for tau isolation and ID #15778 
https://github.com/cms-sw/cmssw/pull/15778	in CMSSW_8_1_X
+1,775 −225	in 19
.. 81 + 116 +23+607+1775 = 2602 is still smaller than 3083 in this 80X PR

```

please check and update the list of already merged 81X/90X PRs in a way that it's clear this is actually a 1:1 backport.

If you plan additional commits on top of this which will be a clear backports of features already in 80X/90X, feel free to add them here

**Comment by roger-wolf** [comment]
Hm, 

I actually made the backport based in the PR's to 81X -- Ok I'll re-check and let you know asap. 

Cheers,
Roger

File: RecoTauTag/Configuration/test/QCD-validation.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/store/relval/CMSSW_8_1_0_pre11/RelValQCD_Pt_80_120_13/GEN-SIM-RECO/81X_mcRun2_asymptotic_Candidate_2016_08_30_11_31_55-v1/00000/24F576D4-4D74-E611-8898-0025905A48BA.root
```
Comment (slava77) [inline RecoTauTag/Configuration/test/QCD-validation.txt]: this seems inappropriate for 80X.
Please check and update to something based on 80X

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+class PATTauDiscriminationAgainstElectronMVA6 : public PATTauDiscriminationProducerBase  
+{
+ public:
+  explicit PATTauDiscriminationAgainstElectronMVA6(const edm::ParameterSet& cfg)
+    : PATTauDiscriminationProducerBase(cfg),
+      mva_(0),
+      category_output_()
+  {
+    mva_ = new AntiElectronIDMVA6(cfg);
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc]: can we use a unique_ptr here as it was introduced in #16030

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    delete mva_;
+  }
+
+private:
+  bool isInEcalCrack(double) const;
+
+  std::string moduleLabel_;
+
+  AntiElectronIDMVA6* mva_;
+  float* mvaInput_;
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc]: what is this needed for here? it looks unused

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+      mvaInput_[6]  = tauDecayMode;
+      mvaInput_[7]  = std::min((float)30., nPhoton);
+      mvaInput_[8]  = std::min((float)0.5, ptWeightedDetaStrip);
+      mvaInput_[9]  = std::min((float)0.5, ptWeightedDphiStrip);
+      mvaInput_[10] = std::min((float)0.5, ptWeightedDrSignal);
+      mvaInput_[11] = std::min((float)0.5, ptWeightedDrIsolation);
+      mvaInput_[12] = std::min((float)100., leadingTrackChi2);
+      mvaInput_[13] = std::min((float)1., eRatio);
+      mvaInput_[14]  = TMath::Sign((float)+1., tau->dxy());
+      mvaInput_[15]  = std::sqrt(std::min((float)1., std::fabs(tau->dxy())));
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc]: here and on L276 the code is different from 90X
```  mvaInput_[15] = std::sqrt(std::fabs(std::min((float)1., std::fabs(tau->dxy()))));```
even though the fabs is redundant, the change should be made in 90X as well for consistency

File: RecoTauTag/Configuration/python/tools/DisplayManager.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    return cv
+
+def checkDifferences(histos):
+    rel_diff = 0
+    if len(histos)>0:
+        for ib in range(0, histos[0].GetNbinsX()):
+            for ih, h in enumerate(histos):
+                if not ih == 0:
+                    if not histos[0].GetBinContent(ib) == 0:
+                        rel_diff+=((h.GetBinContent(ib)-histos[0].GetBinContent(ib))/histos[0].GetBinContent(ib))*((h.GetBinContent(ib)-histos[0].GetBinContent(ib))/histos[0].GetBinContent(ib))
```
Comment (slava77) [inline RecoTauTag/Configuration/python/tools/DisplayManager.py]: you may want to apply the bugfix from https://github.com/cms-sw/cmssw/commit/8ebac5517863086c98e55e3092f13ef0cb6be289

```diff
             for ih, h in enumerate(histos):
                 if not ih == 0:
-                    if not histos[0].GetBinContent(ib) == 0:
-                        rel_diff+=((h.GetBinContent(ib)-histos[0].GetBinContent(ib))/histos[0].GetBinContent(ib))*((h.GetBinContent(ib)-histos[0].GetBinContent(ib))/histos[0].GetBinContent(ib))
+                    if not histos[0].GetBinContent(ib+1) == 0:
+                        rel_diff+=((h.GetBinContent(ib+1)-histos[0].GetBinContent(ib+1))/histos[0].GetBinContent(ib+1))*((h.GetBinContent(ib+1)-histos[0].GetBinContent(ib+1))/histos[0].GetBinContent(ib+1))
     return math.sqrt(rel_diff)
```

**Comment by slava77** [comment]
@roger-wolf 
this is just a weekly check on the status of this PR, based on the pending/unresolved comments in the review.

File: RecoTauTag/Configuration/test/QCD-validation.txt (hunk 1 above)
Comment (roger-wolf) [inline RecoTauTag/Configuration/test/QCD-validation.txt]: These files are only used for our internal tests. They are anyhow obsolete, once the validation samples are deleted form relval at CERN. But I've added some more intuitive dummy files for the next commit.

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc (hunk 2 above)
Comment (roger-wolf) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc]: Ok -- done for the next commit.

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc (hunk 3 above)
Comment (roger-wolf) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationAgainstElectronMVA6.cc]: has been removed. This is left over garbage that had been collected in 81X and also now in 80X.

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc (hunk 4 above)
Comment (roger-wolf) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc]: Yes agreed! I'll do that.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc (hunk 4 above)
Comment (roger-wolf) [inline RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc]: The required update has been provided with https://github.com/cms-sw/cmssw/pull/17620

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17292	f212de7
- the backport of features needed to run from miniAOD is in line with the description and the follow up comments.  The backport is a combination of contributions from #13884, #15731, #15778, #15995, #16030, #17620. Changes are expected in miniAOD as a result of extension of patTau pfEssential.
- jenkins tests pass and comparisons with baseline show no differences.

Local tests confirm no change in RECO and changes in patTau output products.
I'm putting this on hold, to be not merged until the rereco release is to be built.

**Comment by slava77** [comment]
hold

**Comment by slava77** [comment]
@davidlange6 
why was this merged?
This is nominally for miniAOD v2

**Comment by davidlange6** [comment]
ok, i didn't understand that - should I revert?

**Comment by slava77** [comment]
On 4/6/17 9:09 AM, David Lange wrote:
> ok, i didn't understand that - should I revert?


https://github.com/cms-sw/cmssw/issues/17603#issuecomment-291805776

I think it's more appropriate to revert.

**Pull request merged**